components/tidb_query_vec_executors/src/index_scan_executor.rs

```diff
@@ -141,41 +172,105 @@ impl ScanExecutorImpl for IndexScanExecutorImpl {
-            // datum flag.
-            columns.push(LazyBatchColumn::decoded_with_capacity_and_tp(
-                scan_rows,
-                EvalType::Int,
-            ));
+
+        match self.decode_handle_strategy {
+            NoDecode => {}
+            DecodeIntHandle => {
+                columns.push(LazyBatchColumn::decoded_with_capacity_and_tp(
+                    scan_rows,
+                    EvalType::Int,
+                ));
+            }
+            DecodeCommonHandle => {
+                for _ in self.columns_id_without_handle.len()..columns_len {
+                    columns.push(LazyBatchColumn::raw_with_capacity(scan_rows));
+                }
+            }
         }
 
         assert_eq!(columns.len(), columns_len);
         LazyBatchColumnVec::from(columns)
     }
 
+    // Currently, we have 6 foramts of index value.
+    // Value layout:
+    // +--With Restore Data(for indices on string columns)
+    // |  |
+    // |  +--Non Unique (TailLen = len(PaddingData) + len(Flag), TailLen < 8 always)
+    // |  |  |
+    // |  |  |  Layout: TailLen |      RestoreData  |      PaddingData
+    // |  |  |  Length: 1       | size(RestoreData) | size(paddingData)
+    // |  |  |
+    // |  |  |  The length >= 10 always because of padding.
+    // |  |
+    // |  |
+    // |  +--Unique Common Handle
+    // |  |  |
+    // |  |  |
+    // |  |  |  Layout: 0x00 | CHandle Flag | CHandle Len | CHandle       | RestoreData
+    // |  |  |  Length: 1    | 1            | 2           | size(CHandle) | size(RestoreData)
+    // |  |  |
+    // |  |  |  The length > 10 always because of CHandle size.
+    // |  |
+    // |  |
+    // |  +--Unique Integer Handle (TailLen = len(Handle) + len(Flag), TailLen == 8 || TailLen == 9)
+    // |     |
+    // |     |  Layout: 0x08 |    RestoreData    |  Handle
+    // |     |  Length: 1    | size(RestoreData) |   8
+    // |     |
+    // |     |  The length >= 10 always since size(RestoreData) > 0.
+    // |
+    // |
+    // +--Without Restore Data
+    // |
+    // +--Non Unique
+    // |  |
+    // |  |  Layout: '0'
+    // |  |  Length:  1
+    // |
+    // +--Unique Common Handle
+    // |  |
+    // |  |  Layout: 0x00 | CHandle Flag | CHandle Len | CHandle
+    //    |  Length: 1    | 1            | 2           | size(CHandle)
+    // |
+    // |
+    // +--Unique Integer Handle
+    // |
+    // |  Layout: Handle
+    // |  Length:   8
+    #[inline]
     fn process_kv_pair(
         &mut self,
         key: &[u8],
         value: &[u8],
         columns: &mut LazyBatchColumnVec,
     ) -> Result<()> {
+        use DecodeHandleStrategy::*;
+        check_index_key(key)?;
         if value.len() > MAX_OLD_ENCODED_VALUE_LEN {
-            self.process_kv_pair_new(key, value, columns)
+            if value[0] <= 1 && value[1] == table::INDEX_VALUE_COMMON_HANDLE_FLAG {
+                if self.decode_handle_strategy != DecodeCommonHandle {
```

Comment:
You can swap L264 and L265 to be more efficient for non decode common handle cases